client/matchmaking/admin-matchmaking-times.tsx

```diff
@@ -0,0 +1,690 @@
+        <PastMatchmakingTimes activeTab={activeTab} />
+      </HistoryContainer>
+    </>
+  )
+}
+
+function FutureMatchmakingTimes({
+  activeTab,
+  futureTimesIds,
+  setFutureTimesIds,
+  futureTimesById,
+  setFutureTimesById,
+}: {
+  activeTab: MatchmakingType
+  futureTimesIds: ReadonlySet<number>
+  setFutureTimesIds: (updater: Producer<Set<number>>) => void
+  futureTimesById: ReadonlyMap<number, MatchmakingTimeJson>
+  setFutureTimesById: (updater: Producer<Map<number, MatchmakingTimeJson>>) => void
+}) {
+  const dispatch = useAppDispatch()
+  const snackbarController = useSnackbarController()
+
+  const [hasMoreFutureTimes, setHasMoreFutureTimes] = useState(true)
+  const [isLoadingFutureTimes, setIsLoadingFutureTimes] = useState(false)
+  const [lastError, setLastError] = useState<Error | undefined>(undefined)
+
+  const onLoadMoreFutureTimes = useCallback(
+    (offset: number) => {
+      setIsLoadingFutureTimes(true)
+
+      dispatch(
+        getMatchmakingTimesFuture(activeTab, offset, {
+          onSuccess: data => {
+            data.futureTimes.forEach(time => {
+              setFutureTimesIds(draft => {
+                draft.add(time.id)
+              })
+              setFutureTimesById(draft => {
+                draft.set(time.id, time)
+              })
+            })
+            setHasMoreFutureTimes(data.hasMoreFutureTimes)
+            setIsLoadingFutureTimes(false)
+            setLastError(undefined)
+          },
+          onError: err => {
+            setIsLoadingFutureTimes(false)
+            setLastError(err)
+          },
+        }),
+      )
+    },
+    [activeTab, dispatch, setFutureTimesById, setFutureTimesIds],
+  )
+
+  useEffect(() => {
+    const abortController = new AbortController()
+
+    onLoadMoreFutureTimes(0)
+
+    return () => abortController.abort()
+  }, [onLoadMoreFutureTimes])
+
+  const onDeleteMatchmakingTime = (id: number) => {
+    dispatch(
+      deleteMatchmakingTime(activeTab, id, {
+        onSuccess: () => {
+          setFutureTimesIds(draft => {
+            draft.delete(id)
+          })
+
+          snackbarController.showSnackbar('Matchmaking time deleted')
+        },
+        onError: err => {
+          snackbarController.showSnackbar('Error deleting matchmaking time')
+        },
+      }),
+    )
+  }
+
+  if (lastError) {
+    return <ErrorText>Error retrieving upcoming matchmaking times: {lastError.message}</ErrorText>
+  }
+
+  if (futureTimesIds.size === 0) {
+    return null
+  }
+
+  return (
+    <>
+      <SectionTitle>Upcoming Matchmaking Times</SectionTitle>
+      {hasMoreFutureTimes && (
+        <LoadMoreButton
+          label={isLoadingFutureTimes ? 'Loading...' : 'Load more upcoming times'}
+          color='accent'
+          disabled={isLoadingFutureTimes}
+          onClick={() => onLoadMoreFutureTimes(futureTimesIds.size)}
+        />
+      )}
+      <FutureTimesContainer>
+        {Array.from(futureTimesIds).map(id => {
```

Comment:
Array.from's second parameter is a map function, use it instead of calling map after